packages/client/src/runtime/core/engines/common/types/Events.ts

```diff
@@ -1,5 +1,32 @@
-// EventEmitter represents a platform-agnostic slice of NodeJS.EventEmitter,
-export interface EventEmitter {
-  on(event: string, listener: (...args: any[]) => void): unknown
-  emit(event: string, args?: any): boolean
+export type QueryEventType = 'query'
+
+export type LogEventType = 'info' | 'warn' | 'error'
+
+export type EngineEventType = QueryEventType | LogEventType
+
+export type EngineEvent<E extends EngineEventType> = E extends QueryEventType ? QueryEvent : LogEvent
+
+type EngineEventContravariant<E extends EngineEventType> = [E & QueryEventType] extends [never]
+  ? LogEvent
+  : [E & LogEventType] extends [never]
+  ? QueryEvent
+  : QueryEvent & LogEvent
```

Comment:
Could this have been implemented with overloads?

Reply:
Good point, it could've been. We could even make it more strict and only allow `LogEventType` XOR `QueryEventType` by only declaring

```ts
  emit(event: QueryEventType, payload: QueryEvent): boolean
  emit(event: LogEventType, payload: LogEvent): boolean
```

Also, unrelated but we could replace `boolean` with `void` in the return signature since we don't actually need it, it came from the original `EventEmitter` type declaration.